augur/filter/io.py

```diff
@@ -96,46 +99,75 @@ def constant_factory(value):
         raise AugurError(f"missing or malformed priority scores file {fname}")
 
 
-def write_metadata_based_outputs(input_metadata_path: str, delimiters: Sequence[str],
-                                 id_columns: Sequence[str], output_metadata_path: str,
-                                 output_strains_path: str, ids_to_write: Set[str]):
+def get_cat(file):
+    if file.endswith(".gz"):
+        return which("gzcat")
+    if file.endswith(".xz"):
+        return which("xzcat")
+    else:
+        return which("cat")
+
+
+def write_metadata(input_metadata_path: str, delimiters: Sequence[str],
+                   id_columns: Sequence[str], output_metadata_path: str,
+                   ids_to_write: Set[str]):
     """
-    Write output metadata and/or strains file given input metadata information
+    Write output metadata file given input metadata information
     and a set of IDs to write.
     """
     input_metadata = Metadata(input_metadata_path, delimiters, id_columns)
 
-    # Handle all outputs with one pass of metadata. This requires using
-    # conditionals both outside of and inside the loop through metadata rows.
-
-    # Make these conditionally set variables available at this scope.
-    output_metadata_handle = None
-    output_metadata = None
-    output_strains = None
-
-    # Set up output streams.
-    if output_metadata_path:
-        output_metadata_handle = xopen(output_metadata_path, "w")
-        output_metadata = csv.DictWriter(output_metadata_handle, fieldnames=input_metadata.columns,
-                                         delimiter="\t", lineterminator=os.linesep)
-        output_metadata.writeheader()
-    if output_strains_path:
-        output_strains = open(output_strains_path, "w")
-
-    # Write outputs based on rows in the original metadata.
-    for row in input_metadata.rows():
-        row_id = row[input_metadata.id_column]
-        if row_id in ids_to_write:
-            if output_metadata:
-                output_metadata.writerow(row)
-            if output_strains:
-                output_strains.write(row_id + '\n')
-
-    # Close file handles.
-    if output_metadata_handle:
-        output_metadata_handle.close()
-    if output_strains:
-        output_strains.close()
+    tsv_join = which("tsv-join")
```

Comment:
### Using tsv-utils/tsv-join in Augur

@tsibley and I chatted about this yesterday. Two options:

1. Detect `tsv-join` in the environment and use it if available. Otherwise, fall back to the Python approach. Maintenance and additional testing on both code paths would be necessary in this case. This is effectively the same approach as current invocation of `fasttree`/`raxml`/`iqtree`/`vcftools`/etc. except those are explicitly requested by the user while `tsv-join` could be detected and used automatically as a faster alternative to the Python approach.
2. We could bundle `tsv-join` as part of Augur to avoid the the downsides of (1). Based on the latest release [v2.2.1](https://github.com/eBay/tsv-utils/releases/tag/v2.2.1), I thought tsv-utils only distributed binaries for macOS, but it looks like previous versions distribute binaries for both Linux and macOS (and this is how it's [advertised](https://github.com/eBay/tsv-utils/blob/38ed0a1c31742bd8b59196517e89ff0b51e8fb80/README.md#prebuilt-binaries)). I think we can get away with using an older version.

Reply:
> We could bundle tsv-join as part of Augur to avoid the the downsides of (1). 

Last I checked tsv-utils wasn't available for osx-arm64. It may be something we could fix.

Reply:
@victorlin This is a clever solution and the speed-up you observe with ncov data suggests it's worth pursuing! Regarding:

> We could bundle tsv-join as part of Augur to avoid the the downsides of (1).

This seems like the best way to provide this better experience to the most users and follows the pattern of bundling other third-party tools like you mention above.

At first, I liked the idea of tsv-utils being an implementation detail that users don't have to know about, but I wonder about the user experience for people who don't have tsv-utils installed and don't realize why the same command runs slower than in an environment where tsv-utils is available. What if we provided some warning when tsv-utils isn't available to alert users that we are using the fallback implementation? Is there a potential cost to exposing the implementation detail that outweighs the benefit of letting users know they could speed up their filters by installing tsv-utils?

Reply:
> [bundling] seems like the best way to provide this better experience to the most users

I'm wary of the extra work required to figure out how to properly bundle tsv-join with Augur. I'd argue that the best way to provide this experience is already accomplished by including `tsv-join` in the managed runtimes.

> [bundling] follows the pattern of bundling other third-party tools like you mention above.

Oh, I meant that we don't bundle any other third-party tools currently so this would be a new approach.

> What if we provided some warning when tsv-utils isn't available to alert users that we are using the fallback implementation? Is there a potential cost to exposing the implementation detail that outweighs the benefit of letting users know they could speed up their filters by installing tsv-utils?

Great point - I think this will be the easiest way to push the feature through:

- don't bundle
- use `tsv-join` if it's available
- use Python fallback with a warning to consider downloading `tsv-join` in the environment if experiencing slowness

Reply:
> Last I checked tsv-utils wasn't available for osx-arm64. It may be something we could fix.

Cornelius has made this available in [conda-forge](https://anaconda.org/conda-forge/tsv-utils). Note that [bioconda's tsv-utils](https://anaconda.org/bioconda/tsv-utils) still does not support osx-arm64.

Reply:
All bioconda environments always use conda-forge preferentially (if correctly configured) so the migration from bioconda -> conda-forge is not an issue. `conda-base` uses the conda-forge one seamlessly.

Reply:
tsv-utils is built from source over at conda-forge, so it's available for more platforms than the pre-built binaries. linux-aarch64 and osx-arm64 don't have pre-built binaries, but conda-forge has them now.